src/pyfvtool/cell.py

```diff
@@ -240,43 +209,224 @@ def __abs__(self):
+            phi0 = np.copy(self.value)
+            phi0[:, 0] = 0.5*(phi0[:, 0]+phi0[:, 1])
+            phi0[0, :] = 0.5*(phi0[0, :]+phi0[1, :])
+            phi0[:, -1] = 0.5*(phi0[:, -1]+phi0[:, -2])
+            phi0[-1, :] = 0.5*(phi0[-1, :]+phi0[-2, :])
+            phi0[0, 0] = phi0[0, 1]
+            phi0[0, -1] = phi0[0, -2]
+            phi0[-1, 0] = phi0[-1, 1]
+            phi0[-1, -1] = phi0[-1, -2]
+            return (x, y, phi0)
+        elif isinstance(self.domain, Grid3D):
+            x = np.hstack([self.domain.facecenters._x[0],
+                           self.domain.cellcenters._x,
+                           self.domain.facecenters._x[-1]])[:, np.newaxis, np.newaxis]
+            y = np.hstack([self.domain.facecenters._y[0],
+                           self.domain.cellcenters._y,
+                           self.domain.facecenters._y[-1]])[np.newaxis, :, np.newaxis]
+            z = np.hstack([self.domain.facecenters._z[0],
+                           self.domain.cellcenters._z,
+                           self.domain.facecenters._z[-1]])[np.newaxis, np.newaxis, :]
+            phi0 = np.copy(self.value)
+            phi0[:,0,:]=0.5*(phi0[:,0,:]+phi0[:,1,:])
+            phi0[:,-1,:]=0.5*(phi0[:,-2,:]+phi0[:,-1,:])
+            phi0[:,:,0]=0.5*(phi0[:,:,0]+phi0[:,:,0])
+            phi0[:,:,-1]=0.5*(phi0[:,:,-2]+phi0[:,:,-1])
+            phi0[0,:,:]=0.5*(phi0[1,:,:]+phi0[2,:,:])
+            phi0[-1,:,:]=0.5*(phi0[-2,:,:]+phi0[-1,:,:])
+            return (x, y, z, phi0)
+        else:
+            raise NotImplementedError("plotprofile() not implemented for mesh type '{0:s}'".\
+                            format(self.domain.__class__.__name__))
+    
+    
+    def domainIntegral(self) -> float:
+        """
+        Calculate the finite-volume integral of a CellVariable over entire domain
+        
+        The finite-volume integral over the entire mesh domain gives the total
+        amount of `CellVariable` present in the system. Calculation of this
+        integral is useful for checking conservation of the quantity concerned
+        (in case of 'no-flux' BCs), or for monitoring its evolution due to 
+        exchanges via the boundaries (other BCs) or to the presence of source
+        terms.
+    
+        May later become a built-in method of the CellVariable class, but for now 
+        this implementation as a function is chosen for consistency with FVTool. 
+    
+    
+        Returns
+        -------
+        float
+            Total finite-volume integral over entire domain.
+    
+        """
+        v = self.domain.cellVolumes()
+        c = self.internalCellValues
+        return (v*c).flatten().sum()
+
+
+    def BC2GhostCells(self):
+        """
+        assign the boundary values to the ghost cells 
+        
+        Returns
+        -------
+        CellVariable
+            the new cell variable
+        """
+        phi = self.copy()
+        if issubclass(type(phi.domain), Grid1D):
```

Comment:
Perhaps we can directly call bc_to_ghost after creating a copy?

Reply:
I admit that I'm still a bit puzzled about how and when the boundary conditions and ghost cells are handled and updated exactly. I saw that there is some redundant code between `CellVariable.BC2GhostCells()` and `CellVariable.bc_to_ghost()`, but I do not dare touch, since everything seems to work OK for now.

At a certain moment, I thought about embedding (a reference to) the `BoundaryCondtionsBase` in the `CellVariable` object (i.e., `self.BC = BC` in `CellVariable.__init__()`), so that once defined, it is not necessary anymore to recall them explicitly in the user code.

Reply:
We can consider that later, as everything works for now (with the limited testing that we do...).

Reply:
I think I implemented both `CellVariable.BC2GhostCells()` and `CellVariable.bc_to_ghost()` to have one that change the ghost cell values in place and one that creates a new cell by assigning the ghost cell values when creating a new cell variable with known boundary conditions. It is indeed an excellent idea to have boundary conditions as a property of cell values. It has several benefits specially when working with non/linear systems of equations.